lib/iris/tests/__init__.py

```diff
@@ -695,93 +702,99 @@ def _assert_graphic(self):
                 uris = repo.setdefault(unique_id, [])
                 uris.append(uri)
                 print('Creating image file: {}'.format(hash_fname))
-                os.rename(result_fname, hash_fname)
+                figure.savefig(hash_fname)
                 msg = 'Creating imagerepo entry: {} -> {}'
                 print(msg.format(unique_id, uri))
                 with open(repo_fname, 'wb') as fo:
                     json.dump(repo, codecs.getwriter('utf-8')(fo), indent=4,
                               sort_keys=True)
 
-            sha1 = _save_figure_hash()
+            # TBD: Push this fix to imagehash (done!)
+            # See https://github.com/JohannesBuchner/imagehash/pull/31
+            # Now need this imagehash/master pushed to pypi ...
+            def _hex_to_hash(hexstr, hash_size=_HASH_SIZE):
+                l = []
+                count = hash_size * (hash_size // 4)
+                if len(hexstr) != count:
+                    emsg = 'Expected hex string size of {}.'
+                    raise ValueError(emsg.format(count))
+                for i in range(count // 2):
+                    h = hexstr[i*2:i*2+2]
+                    v = int("0x" + h, 16)
+                    l.append([v & 2**i > 0 for i in range(8)])
+                return imagehash.ImageHash(np.array(l))
+
+            # Calculate the test result perceptual image hash.
+            buffer = io.BytesIO()
+            figure = plt.gcf()
+            figure.savefig(buffer, format='png')
+            buffer.seek(0)
+            phash = imagehash.phash(Image.open(buffer), hash_size=_HASH_SIZE)
 
             if unique_id not in repo:
                 if dev_mode:
                     _create_missing()
                 else:
+                    figure.savefig(result_fname)
                     emsg = 'Missing image test result: {}.'
                     raise ValueError(emsg.format(unique_id))
             else:
                 uris = repo[unique_id]
-                # Cherry-pick the registered expected hashes from the
-                # test case uri/s.
-                expected = [os.path.splitext(os.path.basename(uri))[0]
+                # Create the expected perceptual image hashes from the uris.
+                expected = [_hex_to_hash(os.path.splitext(os.path.basename(uri))[0])
                             for uri in uris]
 
-                if sha1.hexdigest() not in expected:
-                    # This can be an accidental failure, unusual, but it occurs
-                    # https://github.com/SciTools/iris/issues/2195
-                    # retry once, in case it passes second time round.
-                    sha1 = _save_figure_hash()
+                # Calculate the hamming distance vector for the result hash.
+                distances = [e - phash for e in expected]
 
-                if sha1.hexdigest() not in expected:
+                if np.all([hd > tol for hd in distances]):
```

Comment:
The result image fails to meet expectation iff it is not similar to all our registered expected images.